src/util-landlock.c

```diff
@@ -0,0 +1,262 @@
+    (LANDLOCK_ACCESS_FS_WRITE_FILE | LANDLOCK_ACCESS_FS_MAKE_DIR | LANDLOCK_ACCESS_FS_MAKE_REG |   \
+            LANDLOCK_ACCESS_FS_REMOVE_FILE | LANDLOCK_ACCESS_FS_MAKE_SOCK)
+
+struct landlock_ruleset {
+    int fd;
+    struct landlock_ruleset_attr attr;
+};
+
+static inline struct landlock_ruleset* LandlockCreateRuleset()
+{
+    struct landlock_ruleset *ruleset = SCMalloc(sizeof(struct landlock_ruleset));
+
+    if (ruleset == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Can't alloc landlock ruleset");
+        return NULL;
+    }
+
+    ruleset->attr.handled_access_fs =
+                _LANDLOCK_ACCESS_FS_READ | _LANDLOCK_ACCESS_FS_WRITE | LANDLOCK_ACCESS_FS_EXECUTE;
+
+    int abi;
+    abi = landlock_create_ruleset(NULL, 0, LANDLOCK_CREATE_RULESET_VERSION);
+    if (abi < 0) {
+        SCFree(ruleset);
+        return NULL;
+    }
+    if (abi < 2) {
+        ruleset->attr.handled_access_fs &= ~LANDLOCK_ACCESS_FS_REFER;
+    }
+
+    ruleset->fd = landlock_create_ruleset(&ruleset->attr, sizeof(ruleset->attr), 0);
+    if (ruleset->fd < 0) {
+        SCFree(ruleset);
+        ruleset = NULL;
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't create landlock ruleset");
+    }
+    return ruleset;
+}
+
+static inline void LandlockEnforceRuleset(struct landlock_ruleset *ruleset)
+{
+    prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0);
+    if (landlock_restrict_self(ruleset->fd, 0))
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't self restrict: %s", strerror(errno));
+}
+
+static int LandlockSandboxingAddRule(struct landlock_ruleset *ruleset, const char *directory, uint64_t permission)
+{
+    struct landlock_path_beneath_attr path_beneath = {
+        .allowed_access = permission & ruleset->attr.handled_access_fs,
+    };
+
+    int dir_fd = open(directory, O_PATH | O_CLOEXEC | O_DIRECTORY);
+    if (dir_fd != -1)
+        path_beneath.parent_fd = dir_fd;
+    else {
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't open %s", directory);
+        return -1;
+    }
+
+    if (landlock_add_rule(ruleset->fd, LANDLOCK_RULE_PATH_BENEATH, &path_beneath, 0)) {
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't add write rule: %s", strerror(errno));
+        close(dir_fd);
+        return -1;
+    }
+
+    if (dir_fd == -1)
+        close(dir_fd);
+
+    return 0;
+}
+
+static inline void LandlockSandboxingWritePath(struct landlock_ruleset *ruleset, const char *directory)
+{
+    if (LandlockSandboxingAddRule(ruleset, directory, _LANDLOCK_SURI_ACCESS_FS_WRITE) == 0) {
+        SCLogInfo("Added write permission to '%s'", directory);
+    }
+}
+
+static inline void LandlockSandboxingReadPath(struct landlock_ruleset *ruleset, const char *directory)
+{
+    if (LandlockSandboxingAddRule(ruleset, directory, _LANDLOCK_ACCESS_FS_READ) == 0) {
+        SCLogInfo("Added read permission to '%s'", directory);
+    }
+}
+
+void LandlockSandboxing(SCInstance *suri)
```

Comment:
I'd like to hear some thoughts about whether we should try to set the complete policy here, which means we encode logic like runmode, `-S` commandline option presence, etc here, or that we would scatter the policy logic over the actual code handling those things like runmodes and commandline option handling.

To me it feels a bit out of place here, but OTOH it is nice to be able to see it in a single place. Thoughts?

Reply:
I agree on the feeling as I got the same. I thought at first: let's add a registration mechanism for modules to add their read and write directories. But I stop myself than doing this because I wanted the protection to be done early in the run. I was also afraid of loosing track of what was really enabled. With current status, the code handling this part is not that big. But I think that if we want to go more granular (v2 of landlock support) with things like per thread permissions then we will need to have a framework in place.

Reply:
I can live with that for this early phase. I guess going forward we might add some file & dir open wrappers that track the paths and annotate them, e.g.
`SCFileOpenOnce` <- would not get in policy if the file access already happened.
`SCFileOpenOften` <- would get into policy
Based on the file open flags we could put them into read or write, and maybe we'd have some additional logic for special operations like the rename case discussed before.

In our support for libcap-ng we have done a per module logic that is applied centrally, I think this ultimately should be done in a similar way.